File: markovChains.py
#Markov chain text generator
import random
import numpy as np

class MarkovChain:
    def __init__(self, text):
        self.text = text

    def generateMarkovChain(self, length=100, EOS=list('.?!'), maxChunkWords=3, probability=1):
        """Generate a markov chain of words of a given lenth"""
        #Split the corpus text up by spaces
        splitText = [w for w in self.text.replace("\n"," ").split(" ") if w != '']
        #If the corpus text is too short, exit
        if len(splitText) < 1+maxChunkWords:
            raise Exception("Corpus text too short")

        #If their is no punctuation in the corpus text, exit
        if not any(x for x in splitText if x[-1] in EOS):
            raise Exception("No punctuation in the corpus text")

        #Generate a dictionary of words against a list of the list of all the next 2 words
        d = {}
        for count,word in enumerate(splitText[:-maxChunkWords]):
            try:
                #If the word is already in the dictionary
                for i in range(maxChunkWords):
                    d[word][i].append(splitText[count+(i+1)])
            except KeyError:
                #Otherwise, add a new dictionary entry
                d[word] = [[splitText[count+(i+1)]] for i in range(maxChunkWords)]


        #Generate the chain based on the dictionary
        chain = ""
        #Choose a starting word at the beginning of the sentence, if there is a full sentence
        word = random.choice([splitText[i] for i in range(len(splitText)) if splitText[i-1][-1] in EOS])

        #Until the chain is too long, and the sentence has ended
        while len(chain.split(" ")) <= length or word[-1] not in EOS:
            #Use a binomial probability to randomly generate the number of words per chunk
            actualChunkWords = int(np.random.binomial(maxChunkWords-1, probability, 1))+1
            #Chose a random index from the first next word list to pick from
            chosenWordIndex = random.randint(0, len(d[word][0])-1)
            #Generate the chunk of words, of length actualChunkWords, from the chosenWordIndex
            chosenWords = [d[word][i][chosenWordIndex] for i in range(actualChunkWords)]
            #Add the chunk to the chain
            chain += " ".join(chosenWords)+" "
            #Set the last word of the chunk as the key for the next chunk
            word = chosenWords[-1]

        #Return the chain, with a capitalised 1st letter
        return chain[0].upper() + chain[1:-1]

def getText(targetFile):
    """Return a string containing the contents of the target file"""
    with open(targetFile, "rt", encoding="utf-8") as f:
        return f.read()

if __name__=="__main__":
    #Generate the chain
    filename = "../corpus.txt" #str(input("Target file: "))
    text = getText(filename)
    t = MarkovChain(text)
    chain = t.generateMarkovChain(100, list('.?!'), 3, .5)
    print(chain)

File: markovChains.cc
//Edmund Goodman - Creative Commons Attribution-NonCommercial-ShareAlike 2.5
//Markov chain text generator
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <fstream>
#include <sstream>
#include <random>

std::string getFileContents(char targetFile[]) {
  //Read contents of the given file into a string
  std::ifstream file( targetFile );
  std::stringstream buffer;
  buffer << file.rdbuf();
  std::string str = buffer.str();
  return str;
}

std::vector<std::string> splitString(std::string text) {
  //Split the string by spaces, to form a vector
  std::istringstream buf(text);
  std::istream_iterator<std::string> beg(buf), end;
  std::vector<std::string> tokens(beg, end);
  return tokens;
}

int generateRandomNumber(int start, int stop) {
  //Return a random number in the given range
  std::random_device rd;
  std::mt19937 eng(rd());
  std::uniform_int_distribution<> distr(start, stop);
  return distr(eng);
}

bool isEndOfSentence(std::string word) {
  //Return whether word is at the end of a sentence (if it ends with punctuation)
  char EOS[3] = {'.','!','?'};
  for (int i=0; i<3; i++) {
    if (word[-1] == EOS[i]) {
      return true;
    }
  }
  return false;
}


int main() {
  //Set initial variables
  int length = 100;

  //Read the contents of the file into a string
  char targetFile[] = "./corpus.txt";
  std::string text = getFileContents(targetFile);

  //Split the string into individual words my spaces
  std::vector<std::string> splitText = splitString(text);

  //Initialise and associative array to store the word, then all of the next words
  std::map <std::string, std::vector<std::string> > d;
  //Loop over the splitText array
  for(std::vector<std::string>::size_type count=0; count != splitText.size(); count++) {
    //If the associative array doesn't have a key of the current word:
    if (d.count( splitText[count] )) {
      //Initialise the array of all following words, then add it to the associative array
      std::vector<std::string> wordArray;
      wordArray.assign(1,splitText[count+1]);
      d[splitText[count]] = wordArray;
    } else {
      //Add the next word to the array of next words
      d[splitText[count]].push_back(splitText[count+1]);
    }
  }

  //Randomly choose a starting word from the start of a sentence, and start the chain
  std::string word = splitText[generateRandomNumber(0, splitText.size())];
  std::string chain = word + " ";

  //While the chain isn't too long, and there is an ongoing sentence
  while (splitString(chain).size() <= length && !isEndOfSentence(word)) {
    //Pick a new word, add it to the chain, than set the new word as the old word
    std::string nextWord = splitText[generateRandomNumber(0, splitText.size())];
    chain += (nextWord + " ");
    word = nextWord;
  }

  //Print the chain
  std::cout << chain << std::endl;

  return 0;
}
